Issue #12600: Streaming Positive LAG

hawkfish opened the pull request:
Use buffering to support streaming computation of constant positive LAGs and negative LEADs that are at most one vector away. This doesn't fix the "look ahead" problem, but the benchmark shows it is about 5x faster than the non-streaming version.

Commit:
Issue #12600: Streaming Positive LAG

Use buffering to support streaming computation of constant positive LAGs
and negative LEADs that are at most one vector away.
This doesn't fix the "look ahead" problem, but the benchmark shows
it is about 5x faster than the non-streaming version.

Commit:
Issue #12600: Streaming Positive LAG

Add new benchmark.

Mytherin commented on src/execution/operator/aggregate/physical_streaming_window.cpp:
```diff
@@ -118,6 +101,98 @@ class StreamingWindowState : public OperatorState {
+		}
+
+		void Execute(ExecutionContext &context, DataChunk &input, Vector &result) {
+			ExpressionExecutor executor(context.client, *wexpr.children[0]);
```
Constructing an `ExpressionExecutor` involves initializing (thread-local) expression states which can be quite expensive, for example if we are doing a regex operation this involves compiling the regex. Could we keep these around for the duration of the execution?

hawkfish replied:
That added another 10%

Mytherin commented on src/execution/operator/aggregate/physical_streaming_window.cpp:
```diff
@@ -179,6 +292,113 @@ unique_ptr<OperatorState> PhysicalStreamingWindow::GetOperatorState(ExecutionCon
+	}
+
+	// Compute the arguments
+	ExpressionExecutor executor(context.client);
```
Unrelated to this PR (as this is just moving the code) - it would be better to keep these cached as well if possible

Mytherin commented on src/execution/operator/aggregate/physical_streaming_window.cpp:
```diff
@@ -118,6 +101,98 @@ class StreamingWindowState : public OperatorState {
+				VectorOperations::Copy(curr, prev, source_count, source_offset, target_offset);
+			} else {
+				//	Overwrite
+				VectorOperations::Copy(curr, prev, source_count, source_offset, target_offset);
```
Profiling this it seems like almost all of the time is going into these `Copy` methods - in particular the nullmask copying is expensive. Since we only have nulls at the beginning, adding this line seems to speed things up significantly as the nullmask copying can then be skipped:

```cpp
FlatVector::Validity(prev).Reset();
```

We should also at some point look into speeding up the nullmask copying particularly when the offsets are aligned (since then we can copy entire `int64_t` blocks instead of doing bit operations per row).

Mytherin replied:
I also have an optimization for `VectorOperations::Copy` here - https://github.com/duckdb/duckdb/pull/12618 - that would improve the current implementation (although maybe not as much as resetting the validity entirely)

hawkfish replied:
Yes that made it 2x faster!

The problem with the `Reset` is that the default may not be `NULL` - it may be some other constant. I'll add a test for that.

hawkfish replied:
It's not a problem - we overwrite the entire buffer. Makes a small difference.

Mytherin commented on test/sql/window/test_streaming_lead_lag.test:
```diff
@@ -0,0 +1,84 @@
+
+query TT
+EXPLAIN
+SELECT i, LAG(i, 1) OVER() AS i1
```
Could we add some tests for different types? e.g.:

```sql
SELECT lag(COLUMNS(*), 1) OVER (), lead(COLUMNS(*), 1) FROM test_all_types()
```

Mytherin replied:
Maybe a nice test is to ensure that lag and lead round-trip, e.g.:
```sql
statement ok
create table all_types as from test_all_types()


foreach col bool tinyint smallint int bigint hugeint uhugeint utinyint usmallint uint ubigint date time timestamp timestamp_s timestamp_ms timestamp_ns time_tz timestamp_tz float double dec_4_1 dec_9_4 dec_18_6 dec38_10 uuid interval varchar blob bit small_enum medium_enum large_enum int_array double_array date_array timestamp_array timestamptz_array varchar_array nested_int_array struct struct_of_arrays array_of_structs map union fixed_int_array fixed_varchar_array fixed_nested_int_array fixed_nested_varchar_array fixed_struct_array struct_of_fixed_array fixed_array_of_int_list list_of_fixed_int_array

query I
SELECT lead(lag, 1) over ()=lag(lead, 1) over ()
· FROM (SELECT lag("${col}", 1) OVER () AS lag, lead("${col}", 1) OVER () AS lead FROM test_all_types())
‣ QUALIFY row_number() over ()==2;
----
true

```

Mytherin reviewed (commented):
Thanks for the PR! Looks great. It would be great if we could have a look at streaming lead/negative lags too - but that could be done in a follow-up PR as well. Some comments from my side:

Mytherin commented on src/execution/operator/aggregate/physical_streaming_window.cpp:
```diff
@@ -118,6 +101,98 @@ class StreamingWindowState : public OperatorState {
+			executor.ExecuteExpression(input, curr);
+
+			//	Copy prev[MAX_BUFFER-offset, MAX_BUFFER] => result[0, offset]
+			idx_t source_count = MAX_BUFFER;
```
Since the lead/lag are likely to be small (in my experience at least, lead/lag of 1 is the most common) - perhaps we can buffer only the required amount of rows for the lead/lag instead of keeping around and shifting an entire vector at all times?

Commit:
Merge branch 'feature' into streaming-lag

Commit:
Issue #12600: Streaming Positive LAG

Incorporate various PR feedback improvements:

* Cached Executors
* Validity Mask reset
* Small Buffers

Along with the mask copy improvement in VectorOperations::Copy
these reduce the benchmark runtime by another 3x.

hawkfish commented:
Viz of the progressive improvements:

![Streaming Lag](https://github.com/duckdb/duckdb/assets/13156216/92808165-b892-429b-bfd9-f7631cf3a4b6)

Mytherin commented:
Thanks!
